simulator/func_sim/alu.h

```diff
@@ -11,11 +11,83 @@
 #include "traps/trap.h"
 
 #include <infra/macro.h>
+
+#include <array>
 #include <tuple>
 
 template<size_t N, typename T>
 T align_up(T value) { return ((value + ((1ull << N) - 1)) >> N) << N; }
 
+struct Mask {
+    const uint32 left_mask, right_mask;
+};
+
+static std::array<Mask, 4> shuffle_consts = {
+    Mask{ 0x00ff0000, 0x0000ff00},
+    Mask{ 0x0f000f00, 0x00f000f0},
+    Mask{ 0x30303030, 0x0c0c0c0c},
+    Mask{ 0x44444444, 0x22222222},
+};
+
+static uint32 shuffle_stage( uint32 src, size_t index)
+{
+    auto [maskL, maskR] = shuffle_consts.at( index);
+    size_t N = 1 << ( 3 - index);
+    uint32 x = src & ~( maskL | maskR);
+    x |= ( ( src << N) & maskL) | ( ( src >> N) & maskR);
+    return x;
+}
+
+static inline uint32 bit_shuffle( uint32 src1, uint32 src2)
+{
+    uint32 result = src1;
+    uint32 shamt = src2 & 15;
+    for( size_t i = 0; i < 4 ; ++i)
+    {
+        if ( shamt & ( 1 << ( 3 - i)))
+            result = shuffle_stage( result, i);
+    }
+    return result;
+}
+
+static inline uint64 bit_shuffle( uint64 src1, uint64 src2)
+{
+    uint64 result = src1;
+    uint32 shamt = narrow_cast<uint32>( src2 & 31);
+    uint32 left_part = narrow_cast<uint32>( result >> 32);
+    uint32 right_part = narrow_cast<uint32>( result);
+    if( shamt & ( 1 << 4))
+    {
+        uint32 buf = left_part & bitmask<uint32>(16);
+        left_part = ( left_part & ~( bitmask<uint32>(16))) | ( ( right_part >> 16) & bitmask<uint32>(16));
+        right_part = ( right_part & ( bitmask<uint32>(16))) | ( buf << 16);
+    }
+    left_part = bit_shuffle( left_part, shamt & 15);
+    right_part = bit_shuffle( right_part, shamt & 15);
+    result = static_cast<uint64>( left_part);
```

Comment:
For widening casting, we usually use curly brackets syntax. It prevents narrow casting, but allows to widen.

```c++
uint64{ left_part};
```